### ISSUE TITLE
[RFC] Composer package autoloader

### ISSUE THREAD

**Issue by maks-rafalko** [infection/infection#1054]
Since PhpSpec and Codeception adapters are not a part of the `infection/infection` anymore, we need to decide how we are going to load them.

### The issue

Currently, `infection/infection` repository have the following [root required packages](https://github.com/infection/infection/blob/14053ccb745a82378cc3572a5ee0390a9f00e0a4/composer.json#L51):

* `infection/codeception-adapter`
* `infection/phpspec-adapter`

They will be removed from `composer.json`, because these packages are *optional*.

However, @BackEndTea [has a concern](https://github.com/infection/infection/issues/922#issuecomment-568258133) regarding DX.

Previously, a developer only needed to `require infection/infection` package and PhpSpec started to work immediately.

Now, after removing `infection/phpspec-adapter` from the core, we would need to do:

```bash
composer require infection/infection
composer require infection/phpspec-adapter
```

or even worse

```bash
composer require infection/infection

bin/infection --test-framework=phpspec

# it fails since PhpSpec Adapter is not installed

composer require infection/phpspec-adapter

bin/infection --test-framework=phpspec
```

### Suggested Solutions

I have 3 solutions for you to discuss.

#### 1. No Package Autoloader

When the user tries to use a custom Test Framework (e.g. Codeception), and it's not installed, Infection throws an exception with the detailed message, explaining what to do.

```bash
composer require infection/infection

bin/infection --test-framework=codeception

# Infection throws an exception

> You are trying to use Codeception extension which is not installed. Please run:
> `composer require infection/codeception-adapter` to install it

# then user installs it manually

composer require infection/codeception-adapter

# and runs Infection again

bin/infection --test-framework=codeception
```

✅ Easy to implement
🔴 Not very friendly
🔴 User needs to execute at least 3 commands

#### 2. Automatically install packages after `composer require infection/infection`

We can write `post-update` / `post-install` scripts to automatically run `composer require infection/codeception-adapter` after user installs `infection/infection`:

```bash
composer require infection/infection

# here custom scripts detects that there is a `codeception.yaml` present at the root of the package
# and automatically calls `composer require infection/codeception-adapter

# then users runs

bin/infection --test-framework=codeception # works!
```

✅ More friendlier than the first solution
🔴 Codeception adapter will be installed even if the user doesn't use it, but uses PHPUnit only

### 3. Automatically install packages on `bin/infection` run

We can automatically install packages when the custom Test Framework Adapter is requested by the developer, and at the same time asking for it:

```bash
composer require infection/infection

# at this point only PHPUnit bundled adapter is present

bin/infection --test-framework=codeception
# here Infection knows users *wants* to use Codeception, and asks developer permission to install it

> Would you like to automatically install `infection/codeception-adapter? [y]

# users answers `yes`

Installing `infection/codeception-adapter`

# and infection continues without interrupting of the process
```

✅ More friendlier than the first solution
✅ Custom Test Framework is installed *only when needed*, comparing to the second solution
✅ By asking `> Would you like to automatically install `infection/codeception-adapter? [y]` and requiring developer's answer, we are addressing @sanmai's [concern](https://github.com/infection/infection/pull/1052#issuecomment-586788342)

> I would advise against that "package autoloader" because it'll go very far out of way from what a composer package usually does, causing unnecessarily surprise, and will require further ado during the setup stage. It is much more obvious and transparent when a user would explicitly the packages they need, whilst requiring only a two-line mention in the docs, and a two-line check in the code, if any.

---

Personally, I would suggest implementing the third solution. Actually, I've already started it (to check if it's possible) and found a way on how to dynamically tell Composer to autoload newly installed classes without interrupting a process.

What do you think @infection/core?

**Comment by theofidry** [comment]
I'm open for the third one, but I hope the code is not too shady and will work fine with Composer 2.

But there is also another point to discuss: what about the PHAR? I would suggest to ship all the adapters in it

**Comment by maks-rafalko** [comment]
> I'm open for the third one, but I hope the code is not too shady and will work fine with Composer 2.

This is a working solution:

```php
$process = new Process([
    (new ComposerExecutableFinder())->find(),
    'require',
    'infection/codeception-adapter'
]);

// ...

$loader = new ClassLoader();

$map = require 'vendor/composer/autoload_psr4.php';

// this code adds new Codeception mapping from `autoload_psr4.php` to Autoloader
foreach ($map as $namespace => $paths) {
    $loader->setPsr4($namespace, $paths);
}

// after this line, `new CodeceptionAdapter(...)` already works!
$loader->register(false);
```

> But there is also another point to discuss: what about the PHAR? I would suggest to ship all the adapters in it

sure, this is exactly what I suggested in https://github.com/infection/infection/issues/922

**Comment by maks-rafalko** [comment]
Btw, found that we already use this approach in e2e tests running from PHPUnit:

https://github.com/infection/infection/blob/14053ccb745a82378cc3572a5ee0390a9f00e0a4/tests/phpunit/Console/E2ETest.php#L271-L288

**Comment by theofidry** [comment]
does that work fine regardless of the autoloader mode? e.g. with `classmap-authoritative`

**Comment by maks-rafalko** [comment]
> does that work fine regardless of the autoloader mode? e.g. with `classmap-authoritative`

checked, it works

**Comment by sanmai** [comment]
I have to say this is a sensible proposal. My worst fears were not realized.

(As someone who authored that part of E2ETest I can say it should totally work.)

**Comment by maks-rafalko** [comment]
Implemented in #1102
